drivers/interrupt_controller/intc_nxp_irqsteer.c

```diff
@@ -0,0 +1,522 @@
+ *			- for example, if an IP uses INTID 34 you'd write its
+ *			interrupts property as follows (i.MX8MP):
+ *				interrupts = <&master1 2>;
+ *
+ * 4) Notes and comments
+ *	> PLEASE DON'T MISTAKE THE ZEPHYR MULTI-LEVEL INTERRUPT
+ *	ORGANIZATION WITH THE XTENSA ONE. THEY ARE DIFFERENT THINGS.
+ *
+ * [1]: https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#multi-level-interrupt-handling
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/irq.h>
+#include <fsl_irqsteer.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/cache.h>
+
+LOG_MODULE_REGISTER(irqsteer);
+
+/* used for driver binding */
+#define DT_DRV_COMPAT nxp_irqsteer_intc
+
+/* utility macros - used for DTS parsing */
+#define _IRQSTEER_REGISTER_DISPATCHER(node_id)				\
+	IRQ_CONNECT(DT_IRQN(node_id),					\
+		    DT_IRQ(node_id, priority),				\
+		    irqsteer_isr_dispatcher,				\
+		    &dispatchers[DT_REG_ADDR(node_id)],			\
+		    0)
+
+#define _IRQSTEER_DECLARE_DISPATCHER(node_id)				\
+{									\
+	.dev = DEVICE_DT_GET(DT_PARENT(node_id)),			\
+	.master_index = DT_REG_ADDR(node_id),				\
+	.irq = DT_IRQN(node_id),					\
+}
+
+#define IRQSTEER_DECLARE_DISPATCHERS(parent_id)\
+	DT_FOREACH_CHILD_STATUS_OKAY_SEP(parent_id, _IRQSTEER_DECLARE_DISPATCHER, (,))
+
+#define IRQSTEER_REGISTER_DISPATCHERS(parent_id)\
+	DT_FOREACH_CHILD_STATUS_OKAY_SEP(parent_id, _IRQSTEER_REGISTER_DISPATCHER, (;))
+
+struct irqsteer_config {
+	uint32_t regmap_phys;
+	uint32_t regmap_size;
+	struct irqsteer_dispatcher *dispatchers;
+};
+
+struct irqsteer_dispatcher {
+	const struct device *dev;
+	/* which set of interrupts is the dispatcher in charge of? */
+	uint32_t master_index;
+	/* which interrupt line is the dispatcher tied to? */
+	uint32_t irq;
+};
+
+static struct irqsteer_dispatcher dispatchers[] = {
+	IRQSTEER_DECLARE_DISPATCHERS(DT_NODELABEL(irqsteer))
+};
+
+static int to_zephyr_irq(uint32_t regmap, uint32_t irq,
+			 struct irqsteer_dispatcher *dispatcher)
+{
+	int i, idx;
+
+	idx = irq;
+
+	for (i = dispatcher->master_index - 1; i >= 0; i--) {
+		idx -= IRQSTEER_GetMasterIrqCount((IRQSTEER_Type *)regmap, i);
+	}
+
+	return irq_to_level_2(idx) | dispatcher->irq;
+}
+
+static int from_zephyr_irq(uint32_t regmap, uint32_t irq, uint32_t master_index)
+{
+	int i, idx;
+
+	idx = irq;
+
+	for (i = 0; i < master_index; i++) {
+		idx += IRQSTEER_GetMasterIrqCount((IRQSTEER_Type *)regmap, i);
+	}
+
+	return idx;
+}
+
+void z_soc_irq_enable_disable(uint32_t irq, bool enable)
+{
+	uint32_t parent_irq;
+	int i, system_irq, level2_irq;
+	const struct irqsteer_config *cfg;
+
+	if (irq_get_level(irq) == 1) {
+		/* LEVEL 1 interrupts are DSP direct */
+		z_xtensa_irq_enable(XTENSA_IRQ_NUMBER(irq));
+		return;
+	}
+
+	parent_irq = irq_parent_level_2(irq);
+	level2_irq = irq_from_level_2(irq);
+
+	/* find dispatcher responsible for this interrupt */
+	for (i = 0; i < ARRAY_SIZE(dispatchers); i++) {
+		if (dispatchers[i].irq == parent_irq) {
+			cfg = dispatchers[i].dev->config;
```

Comment:
I would go with 

``if (dispatchers[i].irq != parent_irq)
     continue;
 ``
 
 So that you remove a level of identation and make code easier to follow

Reply:
Thanks for the suggestion, fixed this.